### ISSUE TITLE
Getting "no binary output function available for type ulid" using SQLx from Rust

### ISSUE THREAD

**Issue by jlandahl** [pksunkara/pgx_ulid#27]
I just installed pgx_ulid into a Postgres 15 database, and while inserting and viewing in a database tool (DataSpell) works fine, when I try to query from Rust using SQLx I get a "no binary output function available for type ulid" error. I thought it might be due to how I was implementing SQLx's `Encode` and `Decode` traits, but I reworked those to be based on strings and I'm still getting the error. I'll paste some example code shortly.

**Comment by jlandahl** [comment]
Here's a minimal program in Rust that exhibits the error:

```rust
use sqlx::postgres::PgPoolOptions;
use sqlx::Row;

#[tokio::main]
async fn main() -> Result<(), sqlx::Error> {
    let connect = "postgres://foo:bar@localhost/db";
    let pool = PgPoolOptions::new()
        .max_connections(5)
        .connect(connect)
        .await?;

    let rows = sqlx::query("SELECT * from sources")
        .fetch_all(&pool)
        .await?;

    println!(
        "rows: {:?}",
        rows.into_iter()
            .map(|row| row.get::<String, _>("source_id"))
            .collect::<Vec<_>>()
    );

    Ok(())
}
```
And here's the full error:

```
Error: Database(PgDatabaseError { severity: Error, code: "42883", message: "no binary output function available for type ulid", detail: None, hint: None, position: None, where: None, schema: None, table: None, column: None, data_type: None, constraint: None, file: Some("lsyscache.c"), line: Some(2946), routine: Some("getTypeBinaryOutputInfo") })
```

**Comment by jlandahl** [comment]
I noticed in the source code that there are implicit conversions to/from UUIDs / ULIDs, so I found this workaround:

```sql
create table foo (
    foo_id uuid primary key default gen_ulid(),
    name text not null
);
```

In other words, store the ULID as a UUID, retrieve it through SQLx as a UUID, and convert to/from ULID in the application. Not ideal, but it works.

**Comment by pksunkara** [comment]
@workingjubilee Can I get some help from you to understand this issue? I store the type as binary but the `InOutFuncs` use string.

Reading https://github.com/launchbadge/sqlx/issues/1269, Does sqlx need another set of `InOutFuncs` to be in binary? If yes, how to add another impl since I already have one?

**Comment by workingjubilee** [comment]
I believe PGRX simply does not adequately support this case unfortunately: we do not have a way to describe the `typsend` and `typreceive` functions for a type.

**Comment by orefalo** [comment]
I am reading all these issues... I wonder if this extension, which looks really good on paper,  is prod ready?

**Comment by pksunkara** [comment]
I use it in production.

**Comment by orefalo** [comment]
indeed, I've been playing with it and it's neat and handy. thank you.

**Comment by nbari** [comment]
Hi @pksunkara 

> `InOutFuncs`

How are you storing created ULids (when not using the extension to generate them)?

I started to use the extension and found this issue, In my case I am creating the Ulid with:
```rust
let token = Ulid::new();
```
And trying to store it with (sqlx)

```rust
let query = "INSERT INTO tokens (id, client_id) VALUES ($1, $2) RETURNING id";
let insert_token = sqlx::query(query)
     .bind(token.to_string())
     .bind(client_id)
     .fetch_one(&mut *tx)
     .await;
```

But I get this error:

    Failed to insert token into database: error returned from database: column "id" is of type ulid but expression is of type text

**Comment by kelvincesar** [comment]
I'm having the same problem. 

To fix in the meanwhile, I updated my struct "UserData.id" to type String. And during the query applied a cast "id::text":
```rust
      match sqlx::query_as::<_, UserData>(
            r#"
            INSERT INTO users (name, email,) VALUES ($1, $2) RETURNING
            id::text, name, email
            "#,
        )
```

**Comment by markmclauchlan** [comment]
I've just tried this with Entity Framework/.net and I've hit the same issue.

Not that helpful I know but I can get it working with this extension: https://github.com/andrielfn/pg-ulid

**Comment by turulix** [comment]
I also found this: https://github.com/theory/pg-semver/issues/59

Where they have an implementation of how they fixed it.
https://github.com/theory/pg-semver/pull/62/files#diff-559832c760eb72f16e53af98ce54edea5921ebb9e01031defe0028f26c9cc8e0

**Comment by pksunkara** [comment]
Released v0.2.2 that fixes this.
